pkg/infra/kvstore/kvstore.go

```diff
@@ -0,0 +1,65 @@
+package kvstore
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+func init() {
+	registry.RegisterService(&KVStore{})
+}
+
+// Key/value store service
+type KVStore struct {
+	SQLStore *sqlstore.SQLStore `inject:""`
+	backend  KVStoreBackend
+}
+
+func (kv *KVStore) Init() error {
+	// currently we only support a SQL backend
+	kv.backend = &KVStoreSQL{
+		SQLStore: kv.SQLStore,
+	}
+
+	return kv.backend.Init()
+}
+
+func (kv *KVStore) Get(ctx context.Context, orgId int64, namespace string, key string) (string, error) {
+	return kv.backend.Get(ctx, orgId, namespace, key)
+}
+
+func (kv *KVStore) Set(ctx context.Context, orgId int64, namespace string, key string, value string) error {
+	return kv.backend.Set(ctx, orgId, namespace, key, value)
+}
+
+// Returns a kvstore wrapper with fixed orgId and namespace
+func (kv *KVStore) WithNamespace(orgId int64, namespace string) *NamespacedKVStore {
+	return &NamespacedKVStore{
+		kvStore:   kv,
+		orgId:     orgId,
+		namespace: namespace,
```

Comment:
@DanCech thoughts on namespace constraints?   This may be similar to the live channel structure where we try to give a pattern that lets internal grafana, plugins, and datasources all use essentially named channels.  In the end, the channel path actually has orgId as a prefix see also https://github.com/grafana/grafana-plugin-sdk-go/blob/master/live/channel.go#L13

but I may be over thinking this -- for sure good to scope things in a way that plugins can safely use this same store

https://grafana.com/docs/grafana/latest/live/live-channel/#channel-structure

Reply:
Interesting, definitely along the same lines.  I had envisioned the namespace basically combining the scope and namespace concepts you have in the channels, using namespaces like `plugin/<slug>` or `dashboard/<uid>` but hadn't formalized what that structure would look like.  We could adopt the same Scope concept here also if that makes sense to you.

If we're using a "normal" k/v store as a backend in the future we would very likely end up mapping to underlying keys like `<org>/<namespace>/<key>` (or `<org>/<scope>/<namespace>/<key>`)

Reply:
> If we're using a "normal" k/v store as a backend in the future we would very likely end up mapping to underlying keys like <org>/<namespace>/<key> (or <org>/<scope>/<namespace>/<key>)

This exactly matches how we build channel names when passing over WebSocket to distinguish between the same channels in different orgs and in some in-memory data structures at the moment. I.e. `<org>/<scope>/<namespace>/<path>` - `key` is called `path` in Live case.

I don't see immediate benefits in attempt to combine two things into one concept, but have a feeling that maybe k/v storage should be more simplified here: i.e. just org_id and key fields (without namespace which can become a part of key itself) but enforce rules on a key creation so that all keys inherit `<scope>/<namespace>/<path>` pattern? At least this will help a bit to avoid possible key duplication in different parts of Grafana. But maybe over-engineering at this point since no clear benefit? 

Also there may be problems with MySQL when trying to build index like:

```
{Cols: []string{"org_id", "namespace", "key"}, Type: UniqueIndex},
```
Where namespace and key are both varchars 190 - there is a limit on index entry size as far as I remember so that MySQL will report errors like `Specified key was too long`

Reply:
The key length issue is a good point, since we don't want to run into problems.  Anyone running mysql 5.7+ is unlikely to run into a problem unless they have explicitly disabled large indexes but it might be an issue for those on older mysql releases.  That said, there is already an index on `user (login, email)` which are both 190 char utf8mb4 columns (despite not being needed as there are also unique indexes on both `login` and `email`) so anyone with large indexes disabled would not be able to install Grafana AFAIK.

One of the reasons I had separated namespace from key was to be able to allocate 190 characters for the key (and another 190 for the namespace) rather than having a 190 character limit, it was also to make it easier to clean up entries later (eg to remove all entries for a specific namespace) without needing to resort to a LIKE query.